### ISSUE TITLE
Parallel Execution

### ISSUE THREAD

**Issue by jhgg** [dittos/graphqllib#28]
Looks like graph resolution is implemented serially currently. I think that we should devise some form of pluggable parallel execution engine that would support Py 3's future, Twisted's defers, threads, greenlets, etc etc... to fan out work.

**Comment by mpaolini** [comment]
I would like to work on a Python3.5+ implementation using `async` and `await` primitives. I think is the most important "parallel" execution engine around because it is in the standard library, has [lots of support](http://asyncio.org/) and is very nice to work with!

The alternative is working with the tulip-like python 3.4 `yield from` and `@coroutine`, this could run even on previous python versions (including 2.7 using tulip).

I would definitely go for the 3.5-only. It would be **very easy** to do if libgraphql only supported 3.5+... is this possible? Do we really need to support older versions?


Anyways, in python3.5+ asyncio, all the functions in the call chain from the entrypoint down to the function that blocks calling the database have to be marked with `async`. 

This means we need:

- an alternative `async` implementation of all the functions that run application code, from `graphql.core.graphql()`, down until the `graphql.core.execution.resolve_field`
- these implementations can reuse all the the synchronous code from `graphql.core`
- so we need to factor out all the synchronous code from the `graphql.core` functions

As an exmple, the main entrypoint `graphql.core.graphql()` would look like:

```python
async def graphql(schema, request='', root=None, vars=None, operation_name=None):
    source = Source(request, 'GraphQL request')  # request is the query fully loaded into memory, sync is OK
    ast = parse(source)  # synchronous parsing is OK for now
    validation_errors = validate(schema, ast)  # validation can synchronous too
    if validation_errors:
        return ExecutionResult(  # I guess this can be synchronous as well
            data=None,
            errors=validation_errors,
        )
    return await execute(  # this is going to implement the business logic, hitting databases and other stuff, has to be async
        schema,
        root or object(),
        ast,
        operation_name,
        vars or {},
    )
```

after factoring out:

```python
async def graphql(schema, request='', root=None, vars=None, operation_name=None):
    ast, errors_result = graphql.core.utils.parse_and_validate(request, 'GraphQLRequest')
    if error_result:
        return error_result
    return await execute
        schema,
        root or object(),
        ast,
        operation_name,
        vars or {},
    )
```

which is not too bad.

Let me know if this makes any sense, thanks!

**Comment by jhgg** [comment]
Py35 is supported and all tests pass on it. I think that the API needs to be refactorted to support any kind of pluggable parallel execution engine.

**Comment by jhgg** [comment]
If execution is pluggable we can support py35 and py27 concurrently. There's no reason to drop py27.

**Comment by mpaolini** [comment]
definitely, is my proposed funciton/module based thing viable for you? In this case I would start working on it right away

**Comment by jhgg** [comment]
Go for it! Imo

**Comment by jhgg** [comment]
Resolved, #51 is merged.
